File: cal.py
INTEGER, PLUS, MINUS, MULTIPLY, DIV, EOF = 'INTEGER', 'PLUS', 'MINUS', 'MULTIPLY', 'DIV', 'EOF'


class Token():
    def __init__(self, type, value):
        self.type = type
        self.value = value

    def __str__(self):
        return 'Token({type}, {value})'.format(
            type=self.type,
            value=repr(self.value)
        )

    def __repr__(self):
        return self.__str__()


class Interpreter():

    def __init__(self, text):
        self.text = text
        self.position = 0
        self.current_token = None
        self.current_char = self.text[self.position]

    def error(self, error="Error parsing input"):
        raise Exception(error)

    def handlePosition(self):
        self.position += 1
        if self.position > len(self.text) - 1:
            # Indicates end of input
            self.current_char = None
        else:
            self.current_char = self.text[self.position]

    def skip_whitespace(self):
        while self.current_char is not None and self.current_char.isspace():
            self.handlePosition()

    def skip(self):
        self.handlePosition()

    def _alpha(self):
        chars = tuple((chr(i) for i in range(ord('a'), ord('z') + 1))) + tuple((chr(i) for i in range(ord('A'), ord('Z') + 1)))
        if self.current_char in chars:
            return True
        else:
            return False

    def _opr(self):
        if self.current_char in ("+", "-", "/", "*"):
            return True
        return False

    def _num(self):
        result = ""
        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self.handlePosition()
            # Enable floating point number.
            if self.current_char == '.':
                result += self.current_char
                self.handlePosition()
        return float(result)

    def tokenizer(self):
        while self.current_char is not None:
            if self.position == 0 and self._opr():
                self.skip()
                continue
            if self._alpha():
                print("Ap")
                self.skip()
                continue
            if self.current_char.isspace():
                self.skip_whitespace()
                continue
            if self.current_char.isdigit():
                return Token(INTEGER, self._num())
            if self.current_char == '+':
                self.handlePosition()
                return Token(PLUS, '+')

            if self.current_char == '-':
                self.handlePosition()
                return Token(MINUS, '-')
            if self.current_char == '*':
                self.handlePosition()
                return Token(MULTIPLY, "*")
            if self.current_char == '/':
                self.handlePosition()
                return Token(DIV, "/")
            self.error()
        return Token(EOF, None)

    def compare(self, type):
        if self.current_token.type == type :
            self.current_token = self.tokenizer()
        else:
            self.error("Operator Error")

    def term(self):
        token = self.current_token
        self.compare(INTEGER)
        return token.value

    def expr(self):
        self.current_token = self.tokenizer()
        result = self.term()
        while self.current_token.type in (PLUS, MINUS, MULTIPLY, DIV):
            if self.position == 1:
                if self._opr():
                    continue
            token = self.current_token
            if token.type == PLUS:
                self.compare(PLUS)
                result = result + self.term()
            elif token.type == MINUS:
                self.compare(MINUS)
                result = result - self.term()
            elif token.type == DIV:
                self.compare(DIV)
                result = result / self.term()
            else:
                self.compare(MULTIPLY)
                result = result * self.term()
        return result


def main():
    while True:
        try:
            text = input('cal-> ')
        except EOFError:
            break
        if not text:
            continue
        interpreter = Interpreter(text)
        result = interpreter.expr()
        # print(interpreter.tokenizer())
        print(result)


if __name__ == "__main__":
    main()

File: cal.cpp
#include <iostream>
#include <string>
#include <cctype>

// Enums for labels.
enum Tokens {INTEGER, PLUS, MINUS, MULTIPLY, DIV,POWER, MODULE, END};
// Python None type.
const char None = '\0';

class Token {
    public:
        Tokens type;
        std::string value;

        Token(){};
        Token(Tokens type, const std::string &value) : type(type), value(value) {}

        // this function behave like magic method in php, python like __str__ , _tostring etc.
        friend std::ostream &operator<<(std::ostream &os, const Token &token) {
            os << "type: " << token.type << " value: " << token.value;
            return os;
        }
};


class Interpreter {
    private:
        Token current_token;
        char current_char;
        int position = 0;
        std::string text;
        long double  result;
    public:
        Interpreter(const std::string &text) {
            this->text = text;
            this->current_char = this->text[this->position];
        }

        // Evaluate the expression.
        Interpreter expr() {
            this->current_token = this->tokenizer();
            float result = this->terms();
            while (this->current_token.type < END) {
                if (this->current_token.type == DIV) {
                    this->compare(DIV);
                    float num = this->terms();
                    if (num == 0) Interpreter::error("Number can not divided by zero");
                    result = result / this->terms();
                }
                if (this->current_token.type == MULTIPLY) {
                    this->compare(MULTIPLY);
                    result = result * this->terms();
                }
                if (this->current_token.type == PLUS) {
                    this->compare(PLUS);
                    result = result + this->terms();
                }
                if (this->current_token.type == MINUS) {
                    this->compare(MINUS);
                    result = result - this->terms();
                }
                if (this->current_token.type == POWER) {
                    this->compare(POWER);
                    double exp = this->terms();
                    while (exp != 1) {
                        result *= result;
                        exp -= 1;
                    }
                }
                if (this->current_token.type == MODULE) {
                    this->compare(MODULE);
                    result = (int) result % (int) this->terms();
                }
            }
            this->result = result;
            return *this;
        }

        // display the output result.
        void print() {
            std::cout << this->result << std::endl;
        }
    private:

        // fire an error.
        static void error(const std::string& error="Error parsing input", bool quit = false) {
            std::cout << error;
            if (quit) exit(1);
        }

        // increment the position to get more characters.
        void handlePosition() {
            this->position += 1;
            if (this->position > (this->text.length() - 1) ) {
                // End of input
                this->current_char = None;
            } else
                this->current_char = this->text[this->position];
        }

        // skip the whitespaces.
        void skipWhitespace() {
            while (this->current_char != None && this->current_char == ' ') {
                this->handlePosition();
            }
        }

        // match the valid operator.
        bool opr() {
            char operators[4] = {'+', '-', '*', '/'};
            for (char opr : operators) {
                if (this->current_char == opr)
                    return true;
            }
            return false;
        }

        //get the number and floating point number.
        float num() {
            std::string result;
            while (this->current_char != None && isdigit(this->current_char)) {
                result += this->current_char;
                this->handlePosition();
                // floating point number
                if (this->current_char == '.') {
                    result += this->current_char;
                    this->handlePosition();
                }
            }
            return std::stof(result);
        }

        //tokenizer
        Token tokenizer() {
            while (this->current_char != None) {
                this->skipWhitespace(); // skip whitespaces.

                if (this->position == 0 && this->opr()) {
                    this->handlePosition(); //skip
                    continue;
                }
                if (this->current_char == '+') {
                    this->handlePosition();
                    return Token(PLUS, "+");
                }
                if (this->current_char == '-') {
                    this->handlePosition();
                    return Token(MINUS, "/");
                }
                if (this->current_char == '*') {
                    this->handlePosition();
                    return Token(MULTIPLY, "*");
                }
                if (this->current_char == '/') {
                    this->handlePosition();
                    return Token(DIV, "/");
                }
                if (this->current_char == '^') {
                    this->handlePosition();
                    return Token(POWER, "^");
                }
                if (this->current_char == '%') {
                    this->handlePosition();
                    return Token(MODULE, "%");
                }
                if (std::isalpha(this->current_char) || (!std::isdigit(this->current_char) && !this->opr())) {
                    this->handlePosition(); // skip
                    Interpreter::error("Alphabets or Special characters are not allowed", true);
                }
                if (isdigit(this->current_char)) {
                    return Token(INTEGER, std::to_string(this->num()));
                }
            }
            return Token(END, "");
        }

        //Compare the token.
        void compare(Tokens T) {
            if (this->current_token.type == T) {
                this->current_token = this->tokenizer();
            } else {
                Interpreter::error("Operator error", true);
            }
        }

        // get the number.
        float terms() {
            Token token = this->current_token;
            this->compare(INTEGER);
            return std::stof(token.value);
        }
};
int main()
{
    std::string input;
    while(true) {
        std::cout << "cal-> ";
        std::getline(std::cin, input);
        Interpreter inpr(input);
        inpr.expr().print();
    }
}
